### DISCUSSION TITLE
`OIKAN v0.0.3`: 🧠 Brainstorm -> Handling non-linear complex relationships beyond the current limitation of polynomial functions [Improve Symbolic Regression]

### DISCUSSION THREAD

**Discussion by silvermete0r** [silvermete0r/oikan#40] [Ideas]
### Current limitation:

- **Polynomial-Only Basis:** The reliance on PolynomialFeatures restricts the model to polynomial relationships, limiting its ability to capture highly non-linear patterns like exponential growth, logarithmic decay, or periodic behavior.
- **Lasso Constraints:** Lasso regression struggles with correlated features, often arbitrarily selecting one from a correlated group, which can reduce interpretability and accuracy.
- **Complexity Explosion:** Higher-degree polynomials increase the feature space exponentially, risking overfitting and computational inefficiency despite Lasso’s sparsity.

### IDEAS:

 - **Replace Lasso with Elastic Net for better feature selection.** 
     - **Why?** Elastic Net combines L1 (Lasso) and L2 (Ridge) regularization, making it better suited for datasets with correlated features. Unlike Lasso, which may select one feature from a correlated group, Elastic Net can select the group, improving stability and interpretability.
     - **Benefits:** Improved handling of multicollinearity, potentially leading to more robust feature selection.
     - **How to implement?**

``` python
from sklearn.linear_model import ElasticNet

model = ElasticNet(alpha=self.alpha, l1_ratio=0.5, fit_intercept=False)
model.fit(X_poly, y)
```
 
 - **Implement hierarchical symbolic regression to efficiently introduce non-linear terms. Two-stage hierarchical approach:**
     - **Stage 1: Coarse Model**
         - Fit a low-degree polynomial model (e.g., degree=2) using Elastic Net to identify main trends and key features.
         - Select the top K features based on coefficient magnitudes (e.g., top_k=5).
     - **Stage 2: Refined Model**
         - For the top K features, generate higher-degree polynomials (e.g., x_i^3) and non-linear transformations (e.g., log(1 + x_i), exp(x_i), sin(x_i)).
         - Combine these with the original low-degree features and fit another Elastic Net model.
     - **Benefits:** Captures non-linear patterns efficiently by focusing computational effort on important features, improving accuracy without excessive complexity.
   
 - **Add SymPy post-processing for simplified, readable formulas.**
     - To enhance human-readability, we can apply symbolic simplification using SymPy after fitting the model;
